src/Launcher.vala

```diff
@@ -82,4 +136,103 @@ public class Dock.Launcher : Gtk.Button {
             return null;
         }
     }
+
+    private Gdk.ContentProvider? on_drag_prepare (double x, double y) {
+        drag_offset_x = (int) x;
+        drag_offset_y = (int) y;
+
+        var val = Value (typeof (Launcher));
+        val.set_object (this);
+        return new Gdk.ContentProvider.for_value (val);
+    }
+
+    private void on_drag_begin (Gtk.DragSource drag_source, Gdk.Drag drag) {
+        var paintable = new Gtk.WidgetPaintable (image); //Maybe TODO How TF can I get a paintable from a gicon?!?!?
+        drag_source.set_icon (paintable.get_current_image (), drag_offset_x, drag_offset_y);
+        image.clear ();
+    }
+
+    private bool on_drag_cancel (Gdk.Drag drag, Gdk.DragCancelReason reason) {
+        if (pinned && reason == NO_TARGET) {
+            pinned = false;
+
+            var settings = new Settings ("io.elementary.dock");
+
+            string[] old_pinned_ids = settings.get_strv ("launchers");
+            string[] new_pinned_ids = {};
+
+            unowned var to_remove_id = app_info.get_id ();
+            foreach (string app_id in old_pinned_ids) {
+                if (app_id != to_remove_id) {
+                    new_pinned_ids += app_id;
+                }
+            }
+
+            settings.set_strv ("launchers", new_pinned_ids);
+
+            var popover = new PoofPopover ();
+
+            unowned var window = (MainWindow) get_root ();
+            popover.set_parent (window);
+            unowned var surface = window.get_surface ();
+
+            double x, y;
+            surface.get_device_position (drag.device, out x, out y, null);
+
+            var rect = Gdk.Rectangle () {
+                x = (int) x,
+                y = (int) y
+            };
+
+            popover.set_pointing_to (rect);
+            // 50 and -13 position the popover in a way that the cursor is in the top left corner.
+            // (TODO: I got this with trial and error and I very much doubt that will be the same everywhere
+            // and at different scalings so it needs testing.)
+            // Although the drag_offset is also measured from the top left corner it works
+            // the other way round (i.e it moves the cursor not the surface)
+            // than set_offset so we put a - in front.
+            popover.set_offset (
+                50 - (drag_offset_x * (popover.width_request / ICON_SIZE)),
+                -13 - (drag_offset_y * (popover.height_request / ICON_SIZE))
+            );
+            popover.popup ();
+            popover.start_animation ();
+
+            var box = (Gtk.Box) parent;
+            if (windows.is_empty ()) {
+                box.remove (this);
+            } else {
+                window.move_launcher_after (this, (Launcher)box.get_last_child ());
+            }
```

Comment:
I think this is pretty tricky. What if we use some model which will keep order in sync with gsettings, so Launcher doesn't have to move itself.

Reply:
Hmm we could do that but I think it would require a pretty big refactor of the current system if done properly so I'd rather do that in another PR. Should it be done before this? Although IMHO it's not that tricky currently but I guess I'm biased here :)

Reply:
I'm ok with adding model later